src/vs/workbench/api/browser/mainThreadNotebookSaveParticipant.ts

```diff
@@ -0,0 +1,67 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { localize } from 'vs/nls';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { IProgressStep, IProgress } from 'vs/platform/progress/common/progress';
+import { extHostCustomer, IExtHostContext } from 'vs/workbench/services/extensions/common/extHostCustomers';
+import { SaveReason } from 'vs/workbench/common/editor';
+import { ExtHostContext, ExtHostNotebookDocumentSaveParticipantShape } from '../common/extHost.protocol';
+import { IDisposable } from 'vs/base/common/lifecycle';
+import { raceCancellationError } from 'vs/base/common/async';
+import { IStoredFileWorkingCopySaveParticipant, IWorkingCopyFileService } from 'vs/workbench/services/workingCopy/common/workingCopyFileService';
+import { IStoredFileWorkingCopy, IStoredFileWorkingCopyModel } from 'vs/workbench/services/workingCopy/common/storedFileWorkingCopy';
+import { NotebookFileWorkingCopyModel } from 'vs/workbench/contrib/notebook/common/notebookEditorModel';
+
+class ExtHostNotebookDocumentSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
+
+	private readonly _proxy: ExtHostNotebookDocumentSaveParticipantShape;
+
+	constructor(extHostContext: IExtHostContext) {
+		this._proxy = extHostContext.getProxy(ExtHostContext.ExtHostNotebookDocumentSaveParticipant);
+	}
+
+	async participate(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, env: { reason: SaveReason }, _progress: IProgress<IProgressStep>, token: CancellationToken): Promise<void> {
+
+		if (!workingCopy.model || !(workingCopy.model instanceof NotebookFileWorkingCopyModel)) {
+			return undefined;
+		}
+
+		const p = new Promise<any>((resolve, reject) => {
+
+			setTimeout(
```

Comment:
seems like this timeout should be cancelled, though I don't know what happens if you reject after resolving. Isn't there a helper method that does this?

Reply:
It would be a noop if the save participant resolves, so the error is not handled then. But yeah it's better to cancel the timeout.